File: XCTestBootstrap/Configuration/FBTestRunnerConfiguration.m
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

#import "FBTestRunnerConfiguration.h"

#import "FBTestConfiguration.h"
#import "FBXCTestConstants.h"
#import "XCTestBootstrapError.h"

@implementation FBTestRunnerConfiguration

#pragma mark Initializers

- (instancetype)initWithSessionIdentifier:(NSUUID *)sessionIdentifier testRunner:(FBBundleDescriptor *)testRunner launchEnvironment:(NSDictionary<NSString *, NSString *> *)launchEnvironment testedApplicationAdditionalEnvironment:(NSDictionary<NSString *, NSString *> *)testedApplicationAdditionalEnvironment testConfiguration:(FBTestConfiguration *)testConfiguration
{
  self = [super init];
  if (!self) {
    return nil;
  }

  _sessionIdentifier = sessionIdentifier;
  _testRunner = testRunner;
  _launchEnvironment = launchEnvironment;
  _testedApplicationAdditionalEnvironment = testedApplicationAdditionalEnvironment;
  _testConfiguration = testConfiguration;

  return self;
}

#pragma mark Public

+ (FBFuture<FBTestRunnerConfiguration *> *)prepareConfigurationWithTarget:(id<FBiOSTarget, FBXCTestExtendedCommands>)target testLaunchConfiguration:(FBTestLaunchConfiguration *)testLaunchConfiguration workingDirectory:(NSString *)workingDirectory codesign:(FBCodesignProvider *)codesign
{
  if (codesign) {
    return [[[codesign
      cdHashForBundleAtPath:testLaunchConfiguration.testBundle.path]
      rephraseFailure:@"Could not determine bundle at path '%@' is codesigned and codesigning is required", testLaunchConfiguration.testBundle.path]
      onQueue:target.asyncQueue fmap:^(id _) {
        return [self prepareConfigurationWithTargetAfterCodesignatureCheck:target testLaunchConfiguration:testLaunchConfiguration workingDirectory:workingDirectory];
      }];
  }
  return [self prepareConfigurationWithTargetAfterCodesignatureCheck:target testLaunchConfiguration:testLaunchConfiguration workingDirectory:workingDirectory];
}

+ (NSDictionary<NSString *, NSString *> *)launchEnvironmentWithHostApplication:(FBBundleDescriptor *)hostApplication hostApplicationAdditionalEnvironment:(NSDictionary<NSString *, NSString *> *)hostApplicationAdditionalEnvironment testBundle:(FBBundleDescriptor *)testBundle testConfigurationPath:(NSString *)testConfigurationPath frameworkSearchPaths:(NSArray<NSString *> *)frameworkSearchPaths
{
  NSMutableDictionary<NSString *, NSString *> *environmentVariables = hostApplicationAdditionalEnvironment.mutableCopy;
  NSString *frameworkSearchPath = [frameworkSearchPaths componentsJoinedByString:@":"];
  [environmentVariables addEntriesFromDictionary:@{
    @"AppTargetLocation" : hostApplication.binary.path,
    @"DYLD_FALLBACK_FRAMEWORK_PATH" : frameworkSearchPath ?: @"",
    @"DYLD_FALLBACK_LIBRARY_PATH" : frameworkSearchPath ?: @"",
    @"OBJC_DISABLE_GC" : @"YES",
    @"TestBundleLocation" : testBundle.path,
    @"XCODE_DBG_XPC_EXCLUSIONS" : @"com.apple.dt.xctestSymbolicator",
    @"XCTestConfigurationFilePath" : testConfigurationPath,
  }];
  return [self addAdditionalEnvironmentVariables:environmentVariables.copy];
}

- (NSArray<NSString *> *)launchArguments
{
  return @[
    @"-NSTreatUnknownArgumentsAsOpen", @"NO",
    @"-ApplePersistenceIgnoreState", @"YES"
  ];
}

#pragma mark NSCopying

- (instancetype)copyWithZone:(NSZone *)zone
{
  return self;
}

#pragma mark Private

+ (NSDictionary<NSString *, NSString *> *)addAdditionalEnvironmentVariables:(NSDictionary<NSString *, NSString *> *)currentEnvironmentVariables
{
  NSString *prefix = @"CUSTOM_";
  NSPredicate *predicate = [NSPredicate predicateWithFormat:@"self BEGINSWITH %@", prefix];
  NSArray *filter = [[NSProcessInfo.processInfo.environment allKeys] filteredArrayUsingPredicate:predicate];
  NSDictionary *envVariableWtihPrefix = [NSProcessInfo.processInfo.environment dictionaryWithValuesForKeys:filter];

  NSMutableDictionary *envs = [currentEnvironmentVariables mutableCopy];
  for (NSString *key in envVariableWtihPrefix)
  {
    envs[[key substringFromIndex:[prefix length]]] = envVariableWtihPrefix[key];
  }

  return [NSDictionary dictionaryWithDictionary:envs];
}


+ (FBFuture<FBTestRunnerConfiguration *> *)prepareConfigurationWithTargetAfterCodesignatureCheck:(id<FBiOSTarget, FBXCTestExtendedCommands>)target testLaunchConfiguration:(FBTestLaunchConfiguration *)testLaunchConfiguration workingDirectory:(NSString *)workingDirectory
{
  // Common Paths
  NSString *runtimeRoot = target.runtimeRootDirectory;
  NSString *platformRoot = target.platformRootDirectory;

  // This directory will contain XCTest.framework, built for the target platform.
  NSString *platformDeveloperFrameworksPath = [platformRoot stringByAppendingPathComponent:@"Developer/Library/Frameworks"];
  // Container directory for XCTest related Frameworks.
  NSString *developerLibraryPath = [runtimeRoot stringByAppendingPathComponent:@"Developer/Library"];
  // Contains other frameworks, depended on by XCTest and Instruments
  NSArray<NSString *> *XCTestFrameworksPaths = @[
    [developerLibraryPath stringByAppendingPathComponent:@"Frameworks"],
    [developerLibraryPath stringByAppendingPathComponent:@"PrivateFrameworks"],
    platformDeveloperFrameworksPath,
  ];

  NSString *automationFrameworkPath = [developerLibraryPath stringByAppendingPathComponent:@"PrivateFrameworks/XCTAutomationSupport.framework"];
  if (![NSFileManager.defaultManager fileExistsAtPath:automationFrameworkPath]) {
    automationFrameworkPath = nil;
  }

  NSMutableDictionary *testedApplicationAdditionalEnvironment = NSMutableDictionary.dictionary;
  NSString *xctTargetBootstrapInjectPath = [platformRoot stringByAppendingPathComponent:@"Developer/usr/lib/libXCTTargetBootstrapInject.dylib"];
  // Xcode > 12.5 does not have this file neither requires it's injection in the target test app.
  if([NSFileManager.defaultManager fileExistsAtPath:xctTargetBootstrapInjectPath]) {
    testedApplicationAdditionalEnvironment[@"DYLD_INSERT_LIBRARIES"] = xctTargetBootstrapInjectPath;
  }

  NSDictionary *testApplicationDependencies = nil;
  if(testLaunchConfiguration.targetApplicationBundle.identifier != nil && testLaunchConfiguration.targetApplicationBundle.path != nil) {
    // Explicitly setting the target application bundle id and path as a dependency triggers XCTest to request
    // the launch of target application via XCTestManager_IDEInterface/XCTMessagingChannel_RunnerToIDE protocols
    testApplicationDependencies = @{
      testLaunchConfiguration.targetApplicationBundle.identifier : testLaunchConfiguration.targetApplicationBundle.path
    };
  }


  // Prepare XCTest bundle
  NSError *error;
  NSUUID *sessionIdentifier = [NSUUID UUID];
  FBBundleDescriptor *testBundle = [FBBundleDescriptor bundleFromPath:testLaunchConfiguration.testBundle.path error:&error];
  if (!testBundle) {
    return [[[XCTestBootstrapError
      describe:@"Failed to prepare test bundle"]
      causedBy:error]
      failFuture];
  }

  // Prepare the test configuration
  FBTestConfiguration *testConfiguration = [FBTestConfiguration
    configurationByWritingToFileWithSessionIdentifier:sessionIdentifier
    moduleName:testBundle.name
    testBundlePath:testBundle.path
    uiTesting:testLaunchConfiguration.shouldInitializeUITesting
    testsToRun:testLaunchConfiguration.testsToRun
    testsToSkip:testLaunchConfiguration.testsToSkip
    targetApplicationPath:testLaunchConfiguration.targetApplicationBundle.path
    targetApplicationBundleID:testLaunchConfiguration.targetApplicationBundle.identifier
    testApplicationDependencies:testApplicationDependencies
    automationFrameworkPath:automationFrameworkPath
    reportActivities:testLaunchConfiguration.reportActivities
    error:&error];
  if (!testConfiguration) {
    return [[[XCTestBootstrapError
      describe:@"Failed to prepare test configuration"]
      causedBy:error]
      failFuture];
  }

  return [[FBFuture
    futureWithFutures:@[
      [target installedApplicationWithBundleID:testLaunchConfiguration.applicationLaunchConfiguration.bundleID],
      [target extendedTestShim],
    ]]
    onQueue:target.asyncQueue map:^(NSArray<id> *tuple) {
      FBInstalledApplication *hostApplication = tuple[0];
      NSString *shimPath = tuple[1];
      NSMutableDictionary<NSString *, NSString *> *hostApplicationAdditionalEnvironment = [NSMutableDictionary dictionary];
      hostApplicationAdditionalEnvironment[kEnv_ShimStartXCTest] = @"1";
      hostApplicationAdditionalEnvironment[@"DYLD_INSERT_LIBRARIES"] = shimPath;
      hostApplicationAdditionalEnvironment[kEnv_WaitForDebugger] = testLaunchConfiguration.applicationLaunchConfiguration.waitForDebugger ? @"YES" : @"NO";
      if (testLaunchConfiguration.coverageDirectoryPath) {
        NSString *continuousCoverageCollectionMode = testLaunchConfiguration.shouldEnableContinuousCoverageCollection ? @"%c" : @"";
        NSString *hostCoverageFile = [NSString stringWithFormat:@"coverage_%@%@.profraw", hostApplication.bundle.identifier, continuousCoverageCollectionMode];
        NSString *hostCoveragePath = [testLaunchConfiguration.coverageDirectoryPath stringByAppendingPathComponent:hostCoverageFile];
        hostApplicationAdditionalEnvironment[kEnv_LLVMProfileFile] = hostCoveragePath;

        if (testLaunchConfiguration.targetApplicationBundle != nil) {
          NSString *targetCoverageFile = [NSString stringWithFormat:@"coverage_%@%@.profraw", testLaunchConfiguration.targetApplicationBundle.identifier, continuousCoverageCollectionMode];
          NSString *targetAppCoveragePath = [testLaunchConfiguration.coverageDirectoryPath stringByAppendingPathComponent:targetCoverageFile];
          testedApplicationAdditionalEnvironment[kEnv_LLVMProfileFile] = targetAppCoveragePath;
        }
      }
      NSString *logDirectoryPath = testLaunchConfiguration.logDirectoryPath;
      if (logDirectoryPath) {
        hostApplicationAdditionalEnvironment[kEnv_LogDirectoryPath] = logDirectoryPath;
      }
      // These Search Paths are added via "DYLD_FALLBACK_FRAMEWORK_PATH" so that they can be resolved when linked by the Application.
      // This is needed so that the Application is aware of how to link the XCTest.framework from the developer directory.
      // The Application binary will not contain linker opcodes that point to the XCTest.framework within the Simulator runtime bundle.
      // Therefore we need to provide them to the test runner so it can pass them to the app launch.
      NSArray<NSString *> *frameworkSearchPaths = [XCTestFrameworksPaths arrayByAddingObject:[hostApplication.bundle.path stringByAppendingPathComponent:@"Frameworks"]];

      // The environment constructed for the app launch must contain the relevant env vars to point at the relevant configuration.
      NSDictionary<NSString *, NSString *> *launchEnvironment = [FBTestRunnerConfiguration
        launchEnvironmentWithHostApplication:hostApplication.bundle
        hostApplicationAdditionalEnvironment:hostApplicationAdditionalEnvironment
        testBundle:testBundle
        testConfigurationPath:testConfiguration.path
        frameworkSearchPaths:frameworkSearchPaths];

      return [[FBTestRunnerConfiguration alloc]
        initWithSessionIdentifier:sessionIdentifier
        testRunner:hostApplication.bundle
        launchEnvironment:launchEnvironment
        testedApplicationAdditionalEnvironment:[testedApplicationAdditionalEnvironment copy]
        testConfiguration:testConfiguration];
    }];
}

@end

File: XCTestBootstrap/Configuration/FBTestRunnerConfiguration.swift
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import Foundation

private let kEnvShimStartXCTest = "SHIMULATOR_START_XCTEST"
private let kEnvWaitForDebugger = "XCTOOL_WAIT_FOR_DEBUGGER"
private let kEnvLLVMProfileFile = "LLVM_PROFILE_FILE"
private let kEnvLogDirectoryPath = "LOG_DIRECTORY_PATH"

/// The ways runner-configuration preparation can fail, as data rather than assembled strings.
public enum FBTestRunnerConfigurationError: Error, LocalizedError {
  case codesignCheckFailed(bundlePath: String, underlying: Error)
  case testBundlePreparationFailed(underlying: Error)
  case testConfigurationPreparationFailed(underlying: Error)

  public var errorDescription: String? {
    switch self {
    case let .codesignCheckFailed(bundlePath, _):
      return "Could not determine bundle at path '\(bundlePath)' is codesigned and codesigning is required"
    case .testBundlePreparationFailed:
      return "Failed to prepare test bundle"
    case .testConfigurationPreparationFailed:
      return "Failed to prepare test configuration"
    }
  }
}

public struct FBTestRunnerConfiguration {

  // MARK: Properties

  public let sessionIdentifier: UUID
  public let testRunner: FBBundleDescriptor
  public let launchEnvironment: [String: String]
  public let testedApplicationAdditionalEnvironment: [String: String]
  public let testConfiguration: FBTestConfiguration

  public var launchArguments: [String] {
    [
      "-NSTreatUnknownArgumentsAsOpen", "NO",
      "-ApplePersistenceIgnoreState", "YES",
    ]
  }

  // MARK: Initializers

  public init(sessionIdentifier: UUID, testRunner: FBBundleDescriptor, launchEnvironment: [String: String], testedApplicationAdditionalEnvironment: [String: String], testConfiguration: FBTestConfiguration) {
    self.sessionIdentifier = sessionIdentifier
    self.testRunner = testRunner
    self.launchEnvironment = launchEnvironment
    self.testedApplicationAdditionalEnvironment = testedApplicationAdditionalEnvironment
    self.testConfiguration = testConfiguration
  }

  // MARK: Public

  public static func prepareConfiguration(withTarget target: FBiOSTarget & ApplicationCommands & XCTestExtendedCommands, testLaunchConfiguration: FBTestLaunchConfiguration, workingDirectory: String, codesign: FBCodesignProvider?) async throws -> FBTestRunnerConfiguration {
    if let codesign {
      do {
        _ = try await bridgeFBFuture(codesign.cdHashForBundle(atPath: testLaunchConfiguration.testBundle.path))
      } catch {
        throw FBTestRunnerConfigurationError.codesignCheckFailed(bundlePath: testLaunchConfiguration.testBundle.path, underlying: error)
      }
    }
    return try await prepareConfigurationAfterCodesignatureCheck(withTarget: target, testLaunchConfiguration: testLaunchConfiguration, workingDirectory: workingDirectory)
  }

  public static func launchEnvironment(withHostApplication hostApplication: FBBundleDescriptor, hostApplicationAdditionalEnvironment: [String: String], testBundle: FBBundleDescriptor, testConfigurationPath: String, frameworkSearchPaths: [String]) -> [String: String] {
    var environmentVariables = hostApplicationAdditionalEnvironment
    let frameworkSearchPath = frameworkSearchPaths.joined(separator: ":")
    environmentVariables["AppTargetLocation"] = hostApplication.binary?.path ?? ""
    environmentVariables["DYLD_FALLBACK_FRAMEWORK_PATH"] = frameworkSearchPath.isEmpty ? "" : frameworkSearchPath
    environmentVariables["DYLD_FALLBACK_LIBRARY_PATH"] = frameworkSearchPath.isEmpty ? "" : frameworkSearchPath
    environmentVariables["OBJC_DISABLE_GC"] = "YES"
    environmentVariables["TestBundleLocation"] = testBundle.path
    environmentVariables["XCODE_DBG_XPC_EXCLUSIONS"] = "com.apple.dt.xctestSymbolicator"
    environmentVariables["XCTestConfigurationFilePath"] = testConfigurationPath
    return addAdditionalEnvironmentVariables(environmentVariables)
  }

  // MARK: Private

  private static func addAdditionalEnvironmentVariables(_ currentEnvironmentVariables: [String: String]) -> [String: String] {
    let prefix = "CUSTOM_"
    var envs = currentEnvironmentVariables
    for (key, value) in ProcessInfo.processInfo.environment {
      if key.hasPrefix(prefix) {
        envs[String(key.dropFirst(prefix.count))] = value
      }
    }
    return envs
  }

  private static func prepareConfigurationAfterCodesignatureCheck(withTarget target: FBiOSTarget & ApplicationCommands & XCTestExtendedCommands, testLaunchConfiguration: FBTestLaunchConfiguration, workingDirectory: String) async throws -> FBTestRunnerConfiguration {
    let runtimeRoot = await target.runtimeRootDirectory
    let platformRoot = await target.platformRootDirectory

    // This directory will contain XCTest.framework, built for the target platform.
    let platformDeveloperFrameworksPath = (platformRoot as NSString).appendingPathComponent("Developer/Library/Frameworks")
    let developerLibraryPath = (runtimeRoot as NSString).appendingPathComponent("Developer/Library")
    // Contains other frameworks, depended on by XCTest and Instruments
    let xcTestFrameworksPaths = [
      (developerLibraryPath as NSString).appendingPathComponent("Frameworks"),
      (developerLibraryPath as NSString).appendingPathComponent("PrivateFrameworks"),
      platformDeveloperFrameworksPath,
    ]

    let automationFrameworkPath = (developerLibraryPath as NSString).appendingPathComponent("PrivateFrameworks/XCTAutomationSupport.framework")
    let automationFrameworkPathOrNil: String? = FileManager.default.fileExists(atPath: automationFrameworkPath) ? automationFrameworkPath : nil

    var testedApplicationAdditionalEnvironment: [String: String] = [:]
    let xctTargetBootstrapInjectPath = (platformRoot as NSString).appendingPathComponent("Developer/usr/lib/libXCTTargetBootstrapInject.dylib")
    // Xcode > 12.5 does not have this file neither requires its injection in the target test app.
    if FileManager.default.fileExists(atPath: xctTargetBootstrapInjectPath) {
      testedApplicationAdditionalEnvironment["DYLD_INSERT_LIBRARIES"] = xctTargetBootstrapInjectPath
    }

    var testApplicationDependencies: [String: String]?
    if let identifier = testLaunchConfiguration.targetApplicationBundle?.identifier,
      let path = testLaunchConfiguration.targetApplicationBundle?.path
    {
      testApplicationDependencies = [identifier: path]
    }

    let sessionIdentifier = UUID()
    let testBundle: FBBundleDescriptor
    do {
      testBundle = try FBBundleDescriptor.bundle(fromPath: testLaunchConfiguration.testBundle.path)
    } catch {
      throw FBTestRunnerConfigurationError.testBundlePreparationFailed(underlying: error)
    }

    let testConfiguration: FBTestConfiguration
    do {
      testConfiguration = try FBTestConfiguration(
        byWritingToFileWithSessionIdentifier: sessionIdentifier,
        moduleName: testBundle.name,
        testBundlePath: testBundle.path,
        uiTesting: testLaunchConfiguration.shouldInitializeUITesting,
        testsToRun: testLaunchConfiguration.testsToRun,
        testsToSkip: testLaunchConfiguration.testsToSkip,
        targetApplicationPath: testLaunchConfiguration.targetApplicationBundle?.path,
        targetApplicationBundleID: testLaunchConfiguration.targetApplicationBundle?.identifier,
        testApplicationDependencies: testApplicationDependencies,
        automationFrameworkPath: automationFrameworkPathOrNil,
        reportActivities: testLaunchConfiguration.reportActivities
      )
    } catch {
      throw FBTestRunnerConfigurationError.testConfigurationPreparationFailed(underlying: error)
    }

    let hostApplication = try await target.installedApplication(bundleID: testLaunchConfiguration.applicationLaunchConfiguration.bundleID)
    let shimPath = try await target.extendedTestShim()

    var hostApplicationAdditionalEnvironment: [String: String] = [:]
    hostApplicationAdditionalEnvironment[kEnvShimStartXCTest] = "1"
    hostApplicationAdditionalEnvironment["DYLD_INSERT_LIBRARIES"] = shimPath
    hostApplicationAdditionalEnvironment[kEnvWaitForDebugger] = testLaunchConfiguration.applicationLaunchConfiguration.waitForDebugger ? "YES" : "NO"

    if let coverageDirectoryPath = testLaunchConfiguration.coverageDirectoryPath {
      let continuousCoverageCollectionMode = testLaunchConfiguration.shouldEnableContinuousCoverageCollection ? "%c" : ""
      let hostCoverageFile = "coverage_\(hostApplication.bundle.identifier)\(continuousCoverageCollectionMode).profraw"
      let hostCoveragePath = (coverageDirectoryPath as NSString).appendingPathComponent(hostCoverageFile)
      hostApplicationAdditionalEnvironment[kEnvLLVMProfileFile] = hostCoveragePath

      if let targetBundle = testLaunchConfiguration.targetApplicationBundle {
        let targetCoverageFile = "coverage_\(targetBundle.identifier)\(continuousCoverageCollectionMode).profraw"
        let targetAppCoveragePath = (coverageDirectoryPath as NSString).appendingPathComponent(targetCoverageFile)
        testedApplicationAdditionalEnvironment[kEnvLLVMProfileFile] = targetAppCoveragePath
      }
    }

    if let logDirectoryPath = testLaunchConfiguration.logDirectoryPath {
      hostApplicationAdditionalEnvironment[kEnvLogDirectoryPath] = logDirectoryPath
    }

    let frameworkSearchPaths = xcTestFrameworksPaths + [(hostApplication.bundle.path as NSString).appendingPathComponent("Frameworks")]

    let launchEnvironment = FBTestRunnerConfiguration.launchEnvironment(
      withHostApplication: hostApplication.bundle,
      hostApplicationAdditionalEnvironment: hostApplicationAdditionalEnvironment,
      testBundle: testBundle,
      testConfigurationPath: testConfiguration.path,
      frameworkSearchPaths: frameworkSearchPaths
    )

    return FBTestRunnerConfiguration(
      sessionIdentifier: sessionIdentifier,
      testRunner: hostApplication.bundle,
      launchEnvironment: launchEnvironment,
      testedApplicationAdditionalEnvironment: testedApplicationAdditionalEnvironment,
      testConfiguration: testConfiguration
    )
  }
}
